### PR TITLE
add the instruction to transfer token internally between accounts own…

### PR DESCRIPTION
**Pull request by Orelsanpls**
add the instruction to transfer token internally between accounts owned by the same governance

### CODE REVIEW

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      mints.map((mint) => tryGetMint(connection.current, mint))
+    )
+
+    return ownedTokenAccountInfos.map((accountInfo) => {
+      const mintDecimals = mintInfos.find((mintInfo) =>
+        mintInfo?.publicKey.equals(accountInfo.mint)
+      )!.account.decimals
+
+      const mintName =
+        Object.values(SPL_TOKENS).find(({ mint }) =>
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: move `Object.values(SPL_TOKENS)` out of the loop?

File: pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  index: number
+  governance: ProgramAccount<Governance> | null
+}) => {
+  const connection = useWalletStore((s) => s.connection)
+  const wallet = useWalletStore((s) => s.current)
+  const { governedMultiTypeAccounts } = useGovernedMultiTypeAccounts()
+
+  // Hardcoded gate used to be clear about what cluster is supported for now
+  if (connection.cluster !== 'mainnet') {
+    return <>This instruction does not support {connection.cluster}</>
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx]: is there a reason for this instruction not to be supported on devnet / testnet ?

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 1 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: Yeah I could enhance mintInfos, good idea

File: pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx (hunk 2 above)
Comment (Orelsanpls) [reply pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx]: Actually no, I've re-enabled it :)

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 1 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: done

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import BigNumber from 'bignumber.js'
+import { useCallback, useEffect, useState } from 'react'
+import { PublicKey } from '@solana/web3.js'
+import { getOwnedTokenAccounts, tryGetMint } from '@utils/tokens'
+import useWalletStore from 'stores/useWalletStore'
+import { SPL_TOKENS } from '@utils/splTokens'
+import { BN } from '@project-serum/anchor'
+import { findATAAddrSync } from '@uxdprotocol/uxd-client'
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts:5]: nit: order imports

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  uiAmount: number
+  mintDecimals: number
+  mintName: string
+
+  // could be either the Associated Token Account of the mint
+  // or a Token Account
+  isATA: boolean
+}
+
+export type OwnedTokenAccountInfos = OwnedTokenAccountInfo[]
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: Useful?

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+// Loads all the token accounts related to the governance public key
+export default function useGovernanceUnderlyingTokenAccounts(
+  governancePk: PublicKey | null
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: Sure need allow it to be `null` ?

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+) {
+  const connection = useWalletStore((state) => state.connection)
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts:79]: retrieving the connection from a `walletStore` mkayyy

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+  const getOwnedTokenAccountsFn = useCallback(async () => {
+    if (!connection || !governancePk) return null
+
+    const accounts = await getOwnedTokenAccounts(
+      connection.current,
+      governancePk
+    )
+
+    const ownedTokenAccountInfos = accounts.map((x) => ({
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: please name things: `x` -> `tokenAccountInfo`

Also `info` is never plural in english, `ownedTokenAccountsInfo`

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      pubkey: x.publicKey,
+      mint: x.account.mint,
+      amount: new BN(x.account.amount.toString()),
+    }))
+
+    // Compute a list of mints related to token accounts
+    const mints = Array.from(
+      ownedTokenAccountInfos.reduce((tmp, x) => tmp.add(x.mint), new Set())
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: looping over `accounts` once, then once again through `ownedTokenAccountInfos` here

Comment (0xCryptoSheik) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: Also, why create a `Set` here ? Many loops wow

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    ) as PublicKey[]
+
+    // load information about the mints
+    const mintInfos = (
+      await Promise.all(
+        mints.map((mint) => tryGetMint(connection.current, mint))
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: looping over the same array again

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+      )
+    ).map((mintInfo, index) => {
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: and once more

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        throw new Error(
+          `Cannot load mint info about ${mints[index].toString()}`
+        )
+      }
+
+      return {
+        mint: mintInfo.publicKey,
+        decimals: mintInfo.account.decimals,
+        name:
+          Object.values(SPL_TOKENS).find(({ mint }) =>
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: so I think this moved a bit but my point was to take `Object.value(SPL_TOKENS)` out of any loops if possible

File: pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx [hunk 12]
```diff
@@ -0,0 +1,202 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import React, { useContext, useEffect, useState } from 'react'
+import * as yup from 'yup'
+import { isFormValid } from '@utils/formValidation'
+import { Token, TOKEN_PROGRAM_ID } from '@solana/spl-token'
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/TokenTransferBetweenInternalGovernanceAccounts.tsx]: nit: cleanup/order imports

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 6 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts:79]: ^^'

That's like it everywhere

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 5 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: you are right, we can use undefined instead with `?`

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 8 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: Set in order to not call `tryGetMint` twice for the same mint

Comment (0xCryptoSheik) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: Maybe extract the mints early, dedupe using Set spread pattern, and then hold a map `{ [mint]: accounts }` ?

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 4 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: This is a map now :)

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 9 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: map now ;)

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 10 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: map now :)

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 11 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: in a function now

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+          amount: new BN(amount.toString()),
+          mint,
+        },
+      }),
+      {} as {
+        [key: string]: Pick<OwnedTokenAccountInfo, 'pubkey' | 'amount' | 'mint'>
+      }
+    )
+
+    const uniqueMintList = Array.from(
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]:
```ts
const uniqueMintList = [
  ...new Set(Object.values(ownedTokenAccountsInfo).map(({ mint }) => mint)),
];
```

File: hooks/useGovernanceUnderlyingTokenAccounts.ts [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+    // Get decimal/name information about the mint
+    const mintsInfo = await getMultipleMintsInfo(
+      uniqueMintList,
+      connection.current
+    )
+
+    // Merge mint info with ownedTokenAccountsInfo
+    return Object.entries(ownedTokenAccountsInfo).reduce(
+      (ownedTokenAccountsInfo, [pubkeyString, ownedTokenAccountInfo]) => {
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceUnderlyingTokenAccounts.ts]: `acc` so it's not confused with the above `ownedTokenAccountsInfo` ?

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  ownedTokenAccountsInfo,
+}: {
+  label: string
+  value?: string
+  onChange: (value: PublicKey) => void
+  error: string
+  ownedTokenAccountsInfo?: OwnedTokenAccountsInfo
+}) => {
+  if (!ownedTokenAccountsInfo) {
+    return <></>
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: `return null;` ?

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  const getAccountDisplay = (pubkey?: PublicKey) => {
+    if (!pubkey) return <div></div>
+
+    const { mint, uiAmount, mintName, isATA } = ownedTokenAccountsInfo[
+      pubkey.toString()
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: `pubkey.toString()` in a variable ?

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  onChange: (value: PublicKey) => void
+  error: string
+  ownedTokenAccountsInfo?: OwnedTokenAccountsInfo
+}) => {
+  if (!ownedTokenAccountsInfo) {
+    return <></>
+  }
+
+  const getAccountDisplay = (pubkey?: PublicKey) => {
+    if (!pubkey) return <div></div>
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: all `ownedTokenAccounts` should have a public key ?

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx (hunk 16 above)
Comment (Orelsanpls) [reply pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: hm yeah ok :)

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx (hunk 17 above)
Comment (Orelsanpls) [reply pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: getAccountDisplay return the data for a given pubkey. If there are no pubkey, then you can't return data. Which is the case when **_value_** is undefined (when nothing is selected)

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 13 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: done

File: hooks/useGovernanceUnderlyingTokenAccounts.ts (hunk 14 above)
Comment (Orelsanpls) [reply hooks/useGovernanceUnderlyingTokenAccounts.ts]: done

File: pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx (hunk 15 above)
Comment (Orelsanpls) [reply pages/dao/[symbol]/proposal/components/instructions/TokenAccountSelect.tsx]: done

**Pull request closed without merging**